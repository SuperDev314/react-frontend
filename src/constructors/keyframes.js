// @flow
import css from './css';
import generateAlphabeticName from '../utils/generateAlphabeticName';
import stringifyRules from '../utils/stringifyRules';
// $FlowFixMe
import hashStr from '../vendor/glamor/hash';
import Keyframes from '../models/Keyframes';

import type { Interpolation, Styles } from '../types';

const replaceWhitespace = (str: string): string => str.replace(/\s|\\n/g, '');

export default function keyframes(
  strings: Styles,
  ...interpolations: Array<Interpolation>
): Keyframes {
  /* Warning if you've used keyframes on React Native */
  if (
    process.env.NODE_ENV !== 'production' &&
    typeof navigator !== 'undefined' &&
    navigator.product === 'ReactNative'
  ) {
    console.warn(
      '`keyframes` cannot be used on ReactNative, only on the web. To do animation in ReactNative please use Animated.'
    );
  }

  const rules = css(strings, ...interpolations);

  const name = generateAlphabeticName(hashStr(replaceWhitespace(JSON.stringify(rules))));

  return new Keyframes(name, stringifyRules(rules, name, '@keyframes'));
}
