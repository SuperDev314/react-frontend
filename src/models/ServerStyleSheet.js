// @flow
/* eslint-disable no-underscore-dangle */
import React from 'react'
import stream from 'stream'

import { IS_BROWSER, SC_STREAM_ATTR } from '../constants'
import StyleSheet from './StyleSheet'
import StyleSheetManager from './StyleSheetManager'

declare var __SERVER__: boolean

/* this error is used for makeStyleTag */
const sheetClosedErr =
  process.env.NODE_ENV !== 'production'
    ? `
Can't collect styles once you've consumed a ServerStyleSheet's styles!
ServerStyleSheet is a one off instance for each server-side render cycle.
- Are you trying to reuse it across renders?
- Are you accidentally calling collectStyles twice?
`.trim()
    : ''

const streamBrowserErr =
  process.env.NODE_ENV !== 'production'
    ? 'Streaming SSR is only supported in a Node.js environment; Please do not try to call this method in the browser.'
    : ''

export default class ServerStyleSheet {
  closed: boolean
  global: StyleSheet
  instance: StyleSheet

  constructor() {
    this.instance = StyleSheet.global.clone()
    this.closed = false
  }

  collectStyles(children: any) {
    if (this.closed) {
      throw new Error(sheetClosedErr)
    }

    return (
      <StyleSheetManager sheet={this.instance}>{children}</StyleSheetManager>
    )
  }

  getStyleTags(): string {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toHTML()
  }

  getStyleElement() {
    if (!this.closed) {
      this.closed = true
    }

    return this.instance.toReactElements()
  }

  interleaveWithNodeStream(readableStream: stream.Readable) {
    if (!__SERVER__ || IS_BROWSER) {
      throw new Error(streamBrowserErr)
    }

    /* the tag index keeps track of which tags have already been emitted */
    const { instance } = this
    let instanceTagIndex = 0

    const streamAttr = `${SC_STREAM_ATTR}="true"`
    const ourStream = new stream.Readable()
    // $FlowFixMe
    ourStream._read = () => {}

    readableStream.on('data', chunk => {
      const { tags } = instance
      let html = ''

      /* retrieve html for each new style tag */
      for (; instanceTagIndex < tags.length; instanceTagIndex += 1) {
        const tag = tags[instanceTagIndex]
        html += tag.toHTML(streamAttr)
      }

      /* force our StyleSheets to emit entirely new tags */
      instance.sealAllTags()
      /* prepend style html to chunk */
      ourStream.push(html + chunk)
    })

    readableStream.on('end', () => {
      this.closed = true
      ourStream.push(null)
    })

    readableStream.on('error', err => {
      this.closed = true
      ourStream.emit('error', err)
    })

    return ourStream
  }
}
