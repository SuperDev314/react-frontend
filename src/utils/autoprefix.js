import camelizeStyleName from 'fbjs/lib/camelizeStyleName'
import hyphenateStyleName from 'fbjs/lib/hyphenateStyleName'
import { autoprefix } from 'glamor/lib/autoprefix'

export default root => {
  root.walkDecls(decl => {
    const objStyle = { [camelizeStyleName(decl.prop)]: decl.value }
    const prefixed = autoprefix(objStyle)
    Object.keys(prefixed).reverse().forEach(newProp => {
      decl.cloneBefore({
        prop: hyphenateStyleName(newProp),
        value: prefixed[newProp],
      })
    })
    decl.remove()
  })
}
