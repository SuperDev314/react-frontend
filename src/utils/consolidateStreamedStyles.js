// @flow
/**
 * When using streaming rendering, style blocks are emitted in chunks directly.
 * This method used to be necessary to move all style elements before the
 * rehydration kicks in. However the rehydration now takes all SSR style tags,
 * combines them into a single tag, and adds it to the head.
 * This renders the consolidation useless
 */
export default function consolidateStreamedStyles() {
  /* noop */
}
