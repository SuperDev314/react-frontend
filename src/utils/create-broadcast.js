/**
 * Creates a broadcast that can be listened to, i.e. simple event emitter
 *
 * @see https://github.com/ReactTraining/react-broadcast
 */
const createBroadcast = (initialValue) => {
  let listeners = []
  let currentValue = initialValue

  return {
    publish(value) {
      currentValue = value
      listeners.forEach(listener => listener(currentValue))
    },
    subscribe(listener) {
      listeners.push(listener)

      // Publish to this subscriber once immediately.
      listener(currentValue)

      // eslint-disable-next-line no-return-assign
      return () =>
        listeners = listeners.filter(item => item !== listener)
    },
  }
}

export default createBroadcast
