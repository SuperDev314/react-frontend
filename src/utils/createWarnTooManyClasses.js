// @flow

import styleSheet from '../models/StyleSheet'

const LIMIT = 200

export default () => {
  let generatedClasses = {}
  let warningSeen = false

  return (className: string, displayName: string) => {
    if (!warningSeen) {
      generatedClasses[className] = true
      if (Object.keys(generatedClasses).length >= LIMIT) {
        const latestRule = styleSheet.rules().find(s => s.selectorText && s.selectorText.indexOf(`.${className}`) !== -1)
        // Unable to find latestRule in test environment.
        /* eslint-disable no-console, prefer-template */
        console.warn(`Over ${LIMIT} classes were generated for component ${displayName}. ` +
           'Consider using style property for frequently changed styles.\n' +
           'Example: const MyStyledComponent = styled(props => <div className={props.className} style={{color: props.color}}>{props.children}</div>)`width: 100px; height: 100px;`' +
           (latestRule ? `\nLatest generated class: ${latestRule.cssText}` : ''))
        warningSeen = true
        generatedClasses = {}
      }
    }
  }
}
