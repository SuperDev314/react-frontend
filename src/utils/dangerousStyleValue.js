// @flow
import unitless from '@emotion/unitless';

// Taken from https://github.com/facebook/react/blob/b87aabdfe1b7461e7331abb3601d9e6bb27544bc/packages/react-dom/src/shared/dangerousStyleValue.js
export default function dangerousStyleValue(name: string, value: any): any {
  // https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/133
  // $FlowFixMe
  if (typeof value === 'symbol') {
    return value;
  }

  const isEmpty = value == null || typeof value === 'boolean' || value === '';
  if (isEmpty) {
    return '';
  }

  if (
    typeof value === 'number' &&
    value !== 0 &&
    !(unitless.hasOwnProperty(name) && unitless[name] === 1)
  ) {
    return `${value}px`; // Presumes implicit 'px' suffix for unitless numbers
  }

  return String(value).trim();
}
