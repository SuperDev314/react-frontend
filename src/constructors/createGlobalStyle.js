// @flow
import React from 'react';
import PropTypes from 'prop-types';
import { IS_BROWSER, STATIC_EXECUTION_CONTEXT } from '../constants';
import GlobalStyle from '../models/GlobalStyle';
import StyleSheet from '../models/StyleSheet';
import { StyleSheetConsumer } from '../models/StyleSheetManager';
import determineTheme from '../utils/determineTheme';
import { ThemeConsumer, type Theme } from '../models/ThemeProvider';
// $FlowFixMe
import hashStr from '../vendor/glamor/hash';
import css from './css';

import type { Interpolation } from '../types';

// place our cache into shared context so it'll persist between HMRs
if (IS_BROWSER) {
  window.scCGSHMRCache = {};
}

export default function createGlobalStyle(
  strings: Array<string>,
  ...interpolations: Array<Interpolation>
) {
  const rules = css(strings, ...interpolations);
  const id = `sc-global-${hashStr(JSON.stringify(rules))}`;
  const style = new GlobalStyle(rules, id);

  class GlobalStyleComponent extends React.Component<*, *> {
    static propTypes = {
      suppressMultiMountWarning: PropTypes.bool,
    };

    static defaultProps = {
      suppressMultiMountWarning: false,
    };

    styleSheet: Object;

    static globalStyle = style;

    static styledComponentId = id;

    constructor() {
      super();

      const { globalStyle, styledComponentId } = this.constructor;

      if (IS_BROWSER) {
        window.scCGSHMRCache[styledComponentId] =
          (window.scCGSHMRCache[styledComponentId] || 0) + 1;
      }

      /**
       * This fixes HMR compatiblility. Don't ask me why, but this combination of
       * caching the closure variables via statics and then persisting the statics in
       * state works across HMR where no other combination did. ¯\_(ツ)_/¯
       */
      this.state = {
        globalStyle,
        styledComponentId,
      };
    }

    componentDidMount() {
      if (
        process.env.NODE_ENV !== 'production' &&
        IS_BROWSER &&
        window.scCGSHMRCache[this.state.styledComponentId] > 1 &&
        !this.props.suppressMultiMountWarning
      ) {
        console.warn(
          `The global style component ${
            this.state.styledComponentId
          } was composed and rendered multiple times in your React component tree. Only the last-rendered copy will have its styles remain in <head> (or your StyleSheetManager target.)`
        );
      }
    }

    componentWillUnmount() {
      if (window.scCGSHMRCache[this.state.styledComponentId]) {
        window.scCGSHMRCache[this.state.styledComponentId] -= 1;
      }
      /**
       * Depending on the order "render" is called this can cause the styles to be lost
       * until the next render pass of the remaining instance, which may
       * not be immediate.
       */
      if (window.scCGSHMRCache[this.state.styledComponentId] === 0) {
        this.state.globalStyle.removeStyles(this.styleSheet);
      }
    }

    render() {
      if (process.env.NODE_ENV !== 'production' && React.Children.count(this.props.children)) {
        console.warn(
          `The global style component ${
            this.state.styledComponentId
          } was given child JSX. createGlobalStyle does not render children.`
        );
      }

      return (
        <StyleSheetConsumer>
          {(styleSheet?: StyleSheet) => {
            this.styleSheet = styleSheet || StyleSheet.master;

            const { globalStyle } = this.state;

            if (globalStyle.isStatic) {
              globalStyle.renderStyles(STATIC_EXECUTION_CONTEXT, this.styleSheet);

              return null;
            } else {
              return (
                <ThemeConsumer>
                  {(theme?: Theme) => {
                    const { defaultProps } = this.constructor;

                    const context = {
                      ...this.props,
                    };

                    if (typeof theme !== 'undefined') {
                      context.theme = determineTheme(this.props, theme, defaultProps);
                    }

                    globalStyle.renderStyles(context, this.styleSheet);

                    return null;
                  }}
                </ThemeConsumer>
              );
            }
          }}
        </StyleSheetConsumer>
      );
    }
  }

  return GlobalStyleComponent;
}
