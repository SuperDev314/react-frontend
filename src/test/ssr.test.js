/**
 * @jest-environment node
 */
// @flow
import React from 'react';
import { renderToString, renderToNodeStream } from 'react-dom/server';
import ServerStyleSheet from '../models/ServerStyleSheet';
import { resetStyled, seedNextClassnames } from './utils';
import keyframes from '../constructors/keyframes';
import createGlobalStyle from '../constructors/createGlobalStyle';

jest.mock('../utils/nonce');

let styled;

describe('ssr', () => {
  beforeEach(() => {
    // eslint-disable-next-line
    require('../utils/nonce').mockReset();

    styled = resetStyled(true);
  });

  afterEach(() => {
    process.env.NODE_ENV = 'test';
  });

  it('should extract the CSS in a simple case', () => {
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(sheet.collectStyles(<Heading>Hello SSR!</Heading>));
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should extract both global and local CSS', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>
      )
    );
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should not spill ServerStyleSheets into each other', () => {
    const A = styled.h1`
      color: red;
    `;
    const B = styled.h1`
      color: green;
    `;

    const sheetA = new ServerStyleSheet();
    renderToString(sheetA.collectStyles(<A />));
    const cssA = sheetA.getStyleTags();

    const sheetB = new ServerStyleSheet();
    renderToString(sheetB.collectStyles(<B />));
    const cssB = sheetB.getStyleTags();

    expect(cssA).toContain('red');
    expect(cssA).not.toContain('green');
    expect(cssB).not.toContain('red');
    expect(cssB).toContain('green');
  });

  it('should add a nonce to the stylesheet if webpack nonce is detected in the global scope', () => {
    // eslint-disable-next-line
    require('../utils/nonce').mockImplementation(() => 'foo');

    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>
      )
    );
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should render CSS in the order the components were defined, not rendered', () => {
    const ONE = styled.h1.withConfig({ componentId: 'ONE' })`
      color: red;
    `;
    const TWO = styled.h2.withConfig({ componentId: 'TWO' })`
      color: blue;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <div>
          <TWO />
          <ONE />
        </div>
      )
    );
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should share global styles but keep renders separate', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const PageOne = styled.h1.withConfig({ componentId: 'PageOne' })`
      color: red;
    `;
    const PageTwo = styled.h2.withConfig({ componentId: 'PageTwo' })`
      color: blue;
    `;

    const sheetOne = new ServerStyleSheet();
    const htmlOne = renderToString(
      sheetOne.collectStyles(
        <React.Fragment>
          <Component />
          <PageOne>Camera One!</PageOne>
        </React.Fragment>
      )
    );
    const cssOne = sheetOne.getStyleTags();

    const sheetTwo = new ServerStyleSheet();
    const htmlTwo = renderToString(
      sheetTwo.collectStyles(
        <React.Fragment>
          <Component />
          <PageTwo>Camera Two!</PageTwo>
        </React.Fragment>
      )
    );
    const cssTwo = sheetTwo.getStyleTags();

    expect(htmlOne).toMatchSnapshot();
    expect(cssOne).toMatchSnapshot();
    expect(htmlTwo).toMatchSnapshot();
    expect(cssTwo).toMatchSnapshot();
  });

  it('should dispatch global styles to each ServerStyleSheet', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Header = styled.h1.withConfig({ componentId: 'Header' })`
      animation: ${props => props.animation} 1s both;
    `;

    seedNextClassnames(['keyframe_0']);

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <React.Fragment>
          <Component />
          <Header animation={keyframes`0% { opacity: 0; }`} />
        </React.Fragment>
      )
    );
    const css = sheet.getStyleTags();

    expect(html).toMatchSnapshot();
    expect(css).toMatchSnapshot();
  });

  it('should return a generated React style element', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <React.Fragment>
          <Component />
          <Heading>Hello SSR!</Heading>
        </React.Fragment>
      )
    );
    const elements = sheet.getStyleElement();

    expect(elements).toHaveLength(1);

    /* I know this looks pointless, but apparently I have the feeling we'll need this */
    expect(elements[0].props.dangerouslySetInnerHTML).toBeDefined();
    expect(elements[0].props.children).not.toBeDefined();

    expect(elements[0].props).toMatchSnapshot();
  });

  it('should return a generated React style element with nonce if webpack nonce is preset in the global scope', () => {
    // eslint-disable-next-line
    require('../utils/nonce').mockImplementation(() => 'foo');

    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const html = renderToString(
      sheet.collectStyles(
        <React.Fragment>
          <Heading>Hello SSR!</Heading>
          <Component />
        </React.Fragment>
      )
    );
    const elements = sheet.getStyleElement();

    expect(elements).toHaveLength(1);
    expect(elements[0].props.nonce).toBe('foo');
  });

  it('should interleave styles with rendered HTML when utilitizing streaming', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(
      <React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
      </React.Fragment>
    );
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        expect(received).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should interleave styles with rendered HTML when chunked streaming', () => {
    const Component = createGlobalStyle`
      body { background: papayawhip; }
    `;
    const Heading = styled.h1`
      color: red;
    `;

    const Body = styled.div`
      color: blue;
    `;

    const SideBar = styled.div`
      color: yellow;
    `;

    const Footer = styled.div`
      color: green;
    `;

    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(
      <React.Fragment>
        <Component />
        <Heading>Hello SSR!</Heading>
        <Body>
          {new Array(1000).fill(0).map((_, i) => (
            <div key={i}>*************************</div>
          ))}
        </Body>
        <SideBar>SideBar</SideBar>
        <Footer>Footer</Footer>
      </React.Fragment>
    );
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      let received = '';

      stream.on('data', chunk => {
        received += chunk;
      });

      stream.on('end', () => {
        expect(received).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        expect(received).toMatch(/yellow/);
        expect(received).toMatch(/green/);
        resolve();
      });

      stream.on('error', reject);
    });
  });

  it('should handle errors while streaming', () => {
    const sheet = new ServerStyleSheet();
    const jsx = sheet.collectStyles(null);
    const stream = sheet.interleaveWithNodeStream(renderToNodeStream(jsx));

    return new Promise((resolve, reject) => {
      stream.on('data', () => {});

      stream.on('error', err => {
        expect(err).toMatchSnapshot();
        expect(sheet.sealed).toBe(true);
        resolve();
      });
    });
  });
});
