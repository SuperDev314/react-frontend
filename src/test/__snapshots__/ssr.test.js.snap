// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssr should return a generated React style element 1`] = `
Object {
  "dangerouslySetInnerHTML": Object {
    "__html": "/* sc-component-id: sc-global-2303210225 */
body{background: papayawhip;}
",
  },
  "data-styled-components": "",
  "data-styled-components-is-local": "false",
  "type": "text/css",
}
`;

exports[`ssr should return a generated React style element 2`] = `
Object {
  "dangerouslySetInnerHTML": Object {
    "__html": "/* sc-component-id: sc-a */
.sc-a {}

.b{color: red;}
",
  },
  "data-styled-components": "b",
  "data-styled-components-is-local": "true",
  "type": "text/css",
}
`;
